frame/contracts/src/exec.rs

```diff
@@ -378,456 +396,640 @@ pub trait Executable<T: Config>: Sized {
-	rent_params: RentParams<T>,
-	_phantom: PhantomData<E>,
+	/// The account seed to be used to instantiate the account counter cache.
+	fn initial_account_seed() -> u64 {
+		<AccountCounter<T>>::get().wrapping_add(1)
+	}
 }
 
-impl<'a, 'b: 'a, T, E> Ext for CallContext<'a, 'b, T, E>
+impl<'a, T, E> Ext for Stack<'a, T, E>
 where
-	T: Config + 'b,
+	T: Config,
 	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
 	E: Executable<T>,
 {
 	type T = T;
 
-	fn get_storage(&self, key: &StorageKey) -> Option<Vec<u8>> {
-		let trie_id = self.ctx.self_trie_id.as_ref().expect(
-			"`ctx.self_trie_id` points to an alive contract within the `CallContext`;\
-				it cannot be `None`;\
-				expect can't fail;\
-				qed",
-		);
-		Storage::<T>::read(trie_id, key)
-	}
-
-	fn set_storage(&mut self, key: StorageKey, value: Option<Vec<u8>>) -> DispatchResult {
-		let trie_id = self.ctx.self_trie_id.as_ref().expect(
-			"`ctx.self_trie_id` points to an alive contract within the `CallContext`;\
-				it cannot be `None`;\
-				expect can't fail;\
-				qed",
-		);
-		// write panics if the passed account is not alive.
-		// the contract must be in the alive state within the `CallContext`;\
-		// the contract cannot be absent in storage;
-		// write cannot return `None`;
-		// qed
-		Storage::<T>::write(&self.ctx.self_account, trie_id, &key, value)
+	fn call(
+		&mut self,
+		gas_limit: Weight,
+		to: T::AccountId,
+		value: BalanceOf<T>,
+		input_data: Vec<u8>,
+	) -> Result<(ExecReturnValue, u32), (ExecError, u32)> {
+		let existing = self
+			.frames()
+			.filter(|f| f.entry_point == ExportedFunction::Call)
+			.find(|f| f.account_id == to)
+			.and_then(|f| {
+				match &f.contract_info {
+					CachedContract::Cached(contract) => Some(contract.clone()),
+					_ => None,
+				}
+			});
+		let executable = self.push_frame(FrameArgs::Call(to, existing), value, gas_limit)?;
+		self.run(executable, input_data)
 	}
 
 	fn instantiate(
 		&mut self,
+		gas_limit: Weight,
 		code_hash: CodeHash<T>,
 		endowment: BalanceOf<T>,
-		gas_meter: &mut GasMeter<T>,
 		input_data: Vec<u8>,
 		salt: &[u8],
 	) -> Result<(AccountIdOf<T>, ExecReturnValue, u32), (ExecError, u32)> {
-		let executable = E::from_storage(code_hash, &self.ctx.schedule, gas_meter)
+		let executable = E::from_storage(code_hash, &self.schedule, self.gas_meter())
 			.map_err(|e| (e.into(), 0))?;
-		let code_len = executable.code_len();
-		self.ctx.instantiate(endowment, gas_meter, executable, input_data, salt)
-			.map(|r| (r.0, r.1, code_len))
-			.map_err(|e| (e, code_len))
-	}
-
-	fn transfer(
-		&mut self,
-		to: &T::AccountId,
-		value: BalanceOf<T>,
-	) -> DispatchResult {
-		transfer::<T>(
-			TransferCause::Call,
-			TransactorKind::Contract,
-			&self.ctx.self_account.clone(),
-			to,
-			value,
-		)
+		let seed = self.next_account_seed();
+		let executable = self.push_frame(
+			FrameArgs::Instantiate(self.frame().account_id.clone(), seed, executable, salt),
+			endowment,
+			gas_limit,
+		)?;
```

Comment:
we don't decrease the account seed in case this fails, I guess it is not important because all we want is an increasing account_seed, so if it doesn't get decreased it is fine, isn't it ?

Reply:
Yeah this is an oversight but it is fine. We only need uniqueness. Sequence is not necessary. We can improve this later.